codegen/options.go

```diff
@@ -1,54 +1,66 @@
 package codegen
 
 import (
+	"github.com/cloudquery/plugin-sdk/schema"
 	"github.com/rs/zerolog"
 )
 
+type TableOption func(*TableDefinition)
+
 // WithNameTransformer overrides how column name will be determined.
-func WithNameTransformer(transformer NameTransformer) TableOptions {
+func WithNameTransformer(transformer NameTransformer) TableOption {
 	return func(t *TableDefinition) {
 		t.nameTransformer = transformer
 	}
 }
 
 // WithSkipFields allows to specify what struct fields should be skipped.
-func WithSkipFields(fields []string) TableOptions {
+func WithSkipFields(fields []string) TableOption {
 	return func(t *TableDefinition) {
 		t.skipFields = fields
 	}
 }
 
 // WithExtraColumns allows passing additional ColumnDefinitions
-func WithExtraColumns(columns ColumnDefinitions) TableOptions {
+func WithExtraColumns(columns ColumnDefinitions) TableOption {
 	return func(t *TableDefinition) {
 		t.extraColumns = columns
 	}
 }
 
-// WithUnwrapFieldsStructs allows to unwrap specific struct fields (1 level deep only)
-func WithUnwrapFieldsStructs(fields []string) TableOptions {
+// WithUnwrapStructFields allows to unwrap specific struct fields (1 level deep only)
+func WithUnwrapStructFields(fields []string) TableOption {
 	return func(t *TableDefinition) {
 		t.structFieldsToUnwrap = fields
 	}
 }
 
 // WithUnwrapAllEmbeddedStructs instructs codegen to unwrap all embedded fields (1 level deep only)
-func WithUnwrapAllEmbeddedStructs() TableOptions {
+func WithUnwrapAllEmbeddedStructs() TableOption {
 	return func(t *TableDefinition) {
 		t.unwrapAllEmbeddedStructFields = true
 	}
 }
 
 // WithLogger allows passing custom logger
-func WithLogger(logger zerolog.Logger) TableOptions {
+func WithLogger(logger zerolog.Logger) TableOption {
 	return func(t *TableDefinition) {
 		t.logger = logger
 	}
 }
 
 // WithTypeTransformer sets a function that can override the schema type for specific fields. Return `schema.TypeInvalid` to fall back to default behavior.
-func WithTypeTransformer(transformer TypeTransformer) TableOptions {
+func WithTypeTransformer(transformer TypeTransformer) TableOption {
 	return func(t *TableDefinition) {
 		t.typeTransformer = transformer
 	}
 }
+
+// WithTypeResolver allows specifying a custom resolver to generate for a specific schema.ValueType.
+// Value should name a schema.ColumnResolverProvider func.
+// Can be specified several times.
+func WithTypeResolver(vt schema.ValueType, resolver string) TableOption {
```

Comment:
I think, when used in conjunction with `WithTypeTransformer`, this will not be flexible enough for some cases, and we need to carefully consider this function signature.

As an example (that I've really seen, but worked around), imagine the struct has a field using a custom timestamp type, called `MyTime`. We can use `WithTypeTransformer` to transform the column type to a normal `time.Time` type. However, we can't directly inject `MyTime` into the column: we need a resolver to convert it from `MyTime` to `time.Time` first.  Now, with the current way `WithTypeResolver` is defined, we would have to generate resolvers for *all* timestamp-ValueType columns, not just `MyTime` ones.

Maybe this is okay, since you could do some type assertions in the resolver to distinguish `time.Time` from `MyTime`? But it seems like there should be a way to define this function so that you can generate a resolver for specific struct fields, rather than specific ValueTypes.

I have no idea if this would be possible to implement, but maybe a signature like this could work:

```
type FieldResolver func(reflect.StructField) (string, error)

func WithFieldResolver(resolver TypeResolver) TableOption {
...
```

WDYT?

Reply:
Right, okay, so you'd have to do a type switch / assertion in the resolver.

I have another toy example I'd like to ask a question about:

```
type MyStruct struct {
    Name string
    Start time.Time
    End time.Time
    Date string
}
```

`Date` is a kind of time, so we would like to use `WithTypeTransformer` to transform it to `schema.TypeTimestamp`. We can look at the name of the field to identify `Date` and do that transformation. Now we also need to parse the date in a resolver before insertion into the column.

With `codegen.WithTypeResolver(schema.TypeTimestamp, "TimestampPathResolver")`, will we be forced to use the same resolver for `Start` and `End` as for `Date`? What if we need a different resolver for `Start` and `End`, or only for `End`?  Can we override it in that case?

Reply:
replaced by `ResolverTransformer` approach